### PR TITLE
[PROF-8804] Add CPU time to wall profiles

### PR DESCRIPTION
**Pull request by nsavoire**
**What does this PR do?**:
Add CPU time measurement as part of wall time profiling samples.
This feature is enabled by a boolean argument in `time-profiler.start` function and is only available on non-windows platform since it requires contexts. Enabling CPU time collection without contexts enabled will result in an error.
For each wall time sample, the current thread CPU since the last sample is measured and attributed to the sample, except if it is an idle sample and in that case CPU will be accounted in the first next non idle sample.

Non-JS thread CPU is reported is a virtual frame named '(non-JS threads)'.
This amount is computed by subtracting the sum of all JS threads CPU from the total process CPU and is only reported in the profile of the main worker JS thread.

**For Datadog employees**:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

**Review by Qard — DISMISSED** [review]

File: bindings/profilers/wall.cc [hunk 1]
```diff
@@ -275,10 +307,16 @@ void SignalHandler::HandleProfilerSignal(int sig,
     return;
   }
 
+  int64_t cpu_time = 0;
+  if (prof->collectCpuTime()) {
+    timespec ts;
+    clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts);
+    cpu_time = timespec_to_duration(ts).count();
```
Comment (szegedi) [inline bindings/profilers/wall.cc:314]: Any reason we shouldn't use 
```suggestion
    cpu_time = currentThreadCpuTime();
```
here? I guess you can mark the method `inline` if you don't want to incur a method call cost for some reason.

File: bindings/profilers/wall.hh [hunk 2]
```diff
@@ -65,7 +65,11 @@ class WallProfiler : public Nan::ObjectWrap {
   bool started_ = false;
   bool workaroundV8Bug_;
   bool detectV8Bug_;
+  bool collectCpuTime_;
+  bool isMainThread_;
   int v8ProfilerStuckEventLoopDetected_ = 0;
+  int64_t startThreadCpuTime_ = 0;
+  int64_t startProcessCpuTime_ = 0;
```
Comment (szegedi) [inline bindings/profilers/wall.hh:72]: `startProcessCpuTime_` looks unused (write-only access) but I presume you'll be using it later.

File: bindings/translate-time-profile.cc [hunk 3]
```diff
@@ -244,9 +247,10 @@ class ProfileTranslator {
...
                                           bool includeLineInfo,
-                                          ContextsByNode* contextsByNode) {
+                                          ContextsByNode* contextsByNode,
+                                          bool hasCpuTime) {
   return ProfileTranslator(contextsByNode)
-      .TranslateTimeProfile(profile, includeLineInfo);
+      .TranslateTimeProfile(profile, includeLineInfo, hasCpuTime);
 }
 
 }  // namespace dd
```
Comment (szegedi) [inline bindings/translate-time-profile.cc:256]: let's add a newline to the end of file

**Review by szegedi — DISMISSED** [review]
This is a very nice, compact addition for a super valuable additional feature. 🚢

File: bindings/profilers/wall.cc (hunk 1 above)
Comment (nsavoire) [reply bindings/profilers/wall.cc]: I might have added `currentThreadCpuTime()` later on and forgot to change all the uses.
Fixed.

**Review by szegedi — CHANGES_REQUESTED** [review]
Overall, amazing. I love how you went through all the trouble of implementing this correctly for Linux, Mach, and Windows (I mean… not that you had a choice, but still appreciate it) and also the thought that went into designing this for correctly computing non-JS CPU by gathering all JS workers times. 👏 

I have just one silly error message fix request 😳

**Review by szegedi — DISMISSED** [review]

File: bindings/profilers/wall.cc [hunk 4]
```diff
@@ -466,8 +520,8 @@ void WallProfiler::Dispose(Isolate* isolate) {
 }
 
 NAN_METHOD(WallProfiler::New) {
-  if (info.Length() != 5) {
-    return Nan::ThrowTypeError("WallProfiler must have four arguments.");
+  if (info.Length() != 7) {
+    return Nan::ThrowTypeError("WallProfiler must have 6 arguments.");
```
[the remark this answers is not in the document]
Comment (Qard) [reply bindings/profilers/wall.cc]: NEED MORE ARGUMENTS! FEED ME ALL THE ARGUMENTS! :goberserk:

File: ts/src/v8-types.ts [hunk 5]
```diff
@@ -59,3 +59,16 @@ export interface Allocation {
...
+  ({node}: {node: AllocationProfileNode}): LabelSet;
+}
+
+export interface GenerateTimeLabelsArgs {
+  node: TimeProfileNode;
+  context?: TimeProfileNodeContext;
+}
+
+export interface GenerateTimeLabelsFunction {
+  ({node, context}: GenerateTimeLabelsArgs): LabelSet;
```
Comment (szegedi) [inline ts/src/v8-types.ts:73]: Just curious: is there an advantage to deconstructing the arg in `{node, context}` over simply `(args: GenerateTimeLabelsArgs): LabelSet`?

**Review by szegedi — APPROVED** [review]
🚢

File: ts/src/v8-types.ts (hunk 5 above)
Comment (nsavoire) [reply ts/src/v8-types.ts]: Hum, no I don't see any advantage of destructuring in the interface.
I think this is just the result of me copy pasting the function args here.
I will remove it.

Comment (nsavoire) [reply ts/src/v8-types.ts]: Done

**Review by szegedi — APPROVED** [review]
🚢 🚢

**Pull request merged**